File: Objective-C/iOS/ParentViewController.m
/*
See LICENSE folder for this sample’s licensing information.

Abstract:
Manages the child view controllers: Products and Purchases. Displays a Restore button that allows you to restore all previously purchased
 non-consumable and auto-renewable subscription products. Requests product information about a list of product identifiers using StoreManager. Calls
 StoreObserver to implement the restoration of purchases.
*/

#import "Products.h"
#import "Purchases.h"
#import "Utilities.h"
#import "StoreManager.h"
#import "StoreObserver.h"
#import "AppConfiguration.h"
#import "ParentViewController.h"

typedef NS_ENUM(NSInteger, ParentViewControllerSegment) {
	ParentViewControllerSegmentProducts = 0,
	ParentViewControllerSegmentPurchases
};

@interface ParentViewController ()
@property (weak, nonatomic) IBOutlet UIView *containerView;
@property (weak, nonatomic) IBOutlet UISegmentedControl *segmentedControl;
@property (weak, nonatomic) IBOutlet UIBarButtonItem *restoreButton;

@property (strong) Products *products;
@property (strong) Purchases *purchases;
@property (strong) Utilities *utility;
@end

@implementation ParentViewController

- (instancetype)initWithCoder:(NSCoder *)coder {
	self = [super initWithCoder:coder];

	if (self != nil) {
		_utility = [[Utilities alloc] init];
	}
	return self;
}

#pragma mark - View Life Cycle

- (void)viewDidLoad {
	[super viewDidLoad];

	// Disable or hide items.
	self.restoreButton.enabled = NO;

	self.products = [self instantiateProductsWithData:[NSMutableArray array]];
	self.purchases = [self instantiatePurchasesWithData:[NSMutableArray array]];

	// Registers for product request and purchase notifications.
	[[NSNotificationCenter defaultCenter] addObserver:self
											 selector:@selector(handleProductRequestNotification:)
												 name:PCSProductRequestNotification
											   object:[StoreManager sharedInstance]];

	[[NSNotificationCenter defaultCenter] addObserver:self
											 selector:@selector(handlePurchaseNotification:)
												 name:PCSPurchaseNotification
											   object:[StoreObserver sharedInstance]];
	// Fetch product information.
	[self fetchProductInformation];
}

#pragma mark - Instantiate View Controllers

-(Products *)instantiateProductsWithData:(NSMutableArray *)data {
	Products *controller = (Products *)[self.storyboard instantiateViewControllerWithIdentifier:PCSViewControllerIdentifiersProducts];
	controller.data = data;
	return controller;
}

-(Purchases *)instantiatePurchasesWithData:(NSMutableArray *)data {
	Purchases *controller = (Purchases *)[self.storyboard instantiateViewControllerWithIdentifier:PCSViewControllerIdentifiersPurchases];
	controller.data = data;
	return controller;
}

#pragma mark - Switching Between View Controllers

/// Adds a child view controller to the container.
-(void)addBaseViewController:(BaseViewController *)viewController {
	[self addChildViewController:viewController];
	viewController.view.translatesAutoresizingMaskIntoConstraints = NO;
	viewController.view.frame = (self.containerView).bounds;
	[self.containerView addSubview:viewController.view];

	[NSLayoutConstraint activateConstraints:@[[viewController.view.topAnchor constraintEqualToAnchor: self.containerView.safeAreaLayoutGuide.topAnchor],
											  [viewController.view.bottomAnchor constraintEqualToAnchor: self.containerView.safeAreaLayoutGuide.bottomAnchor],
											  [viewController.view.leadingAnchor constraintEqualToAnchor: self.containerView.safeAreaLayoutGuide.leadingAnchor],
											  [viewController.view.trailingAnchor constraintEqualToAnchor: self.containerView.safeAreaLayoutGuide.trailingAnchor]]];
	[viewController didMoveToParentViewController:self];
}

/// Removes a child view controller from the container.
-(void)removeBaseViewController:(BaseViewController *)ViewController {
	if (ViewController != nil) {
		[ViewController willMoveToParentViewController:nil];
		[ViewController.view removeFromSuperview];
		[ViewController removeFromParentViewController];
	}
}

/// Switches between the Products and Purchases view controllers.
-(void)switchToViewController:(ParentViewControllerSegment)segment {
	switch(segment){
		case ParentViewControllerSegmentProducts:
			[self removeBaseViewController:self.purchases];
			[self addBaseViewController:self.products];
			break;
		case ParentViewControllerSegmentPurchases:
			[self removeBaseViewController:self.products];
			[self addBaseViewController:self.purchases];
			break;
	}
}

#pragma mark - Fetch Product Information

/// Retrieves product information from the App Store.
-(void)fetchProductInformation {
	// First, let's check whether the user is allowed to make purchases. Proceed if they are allowed. Display an alert, otherwise.
	if ([StoreObserver sharedInstance].isAuthorizedForPayments) {
		self.restoreButton.enabled = YES;
		NSArray *identifiers = self.utility.identifiers;

		if (identifiers != nil) {
			if (identifiers.count > 0) {
				Section *section = [[Section alloc] initWithName:PCSProductsInvalidIdentifiers elements:identifiers];

				// Refresh the UI with identifiers to be queried.
				[self switchToViewController:ParentViewControllerSegmentProducts];
				[self.products reloadWithData:[NSMutableArray arrayWithObject:section]];

				// Fetch the product information.
				[[StoreManager sharedInstance] startProductRequestWithIdentifiers:identifiers];
			} else {
				// Warn the user that the resource file does not contain anything.
				[self alertWithTitle:PCSMessagesStatus message:[NSString stringWithFormat:@"%@.%@ %@", PCSProductIdsPlistName, PCSProductIdsPlistFileExtension, PCSMessagesEmptyResource]];
			}
		} else {
			// Warn the user that the resource file could not be found.
			[self alertWithTitle:PCSMessagesStatus message:[NSString stringWithFormat:@"%@ %@.%@.", PCSMessagesResourceNotFound, PCSProductIdsPlistName, PCSProductIdsPlistFileExtension]];
		}
	} else {
		// Warn the user that they are not allowed to make purchases.
		[self alertWithTitle:PCSMessagesStatus message:[NSString stringWithFormat:@"%@", PCSMessagesCannotMakePayments]];
	}
}

#pragma mark - Display Alert

/// Creates and displays an alert.
-(void)alertWithTitle:(NSString *)title message:(NSString *)message {
	UIAlertController *alertController = [self.utility alertWithTitle:title message:message];
	[self.navigationController presentViewController:alertController animated:YES completion:nil];
}

#pragma mark - Restore All Appropriate Transactions

/// Called when tapping the "Restore" button in the UI.
- (IBAction)restore:(UIBarButtonItem *)sender {
	// Call StoreObserver to restore all restorable purchases.
	[[StoreObserver sharedInstance] restore];
}

#pragma mark - Handle Segmented Control Tap

/// Called when tapping any segmented control in the UI.
- (IBAction)segmentedControlSelectionDidChange:(UISegmentedControl *)sender {
	[self switchToViewController:sender.selectedSegmentIndex];

	switch(sender.selectedSegmentIndex){
		case ParentViewControllerSegmentProducts:
			[self fetchProductInformation];
			break;
		case ParentViewControllerSegmentPurchases:
			[self.purchases reloadWithData:self.utility.dataSourceForPurchasesUI];
			break;
	}
}

#pragma mark - Handle PCSProductRequest Notification

/// Updates the UI according to the product request notification result.
-(void)handleProductRequestNotification:(NSNotification *)notification {
	StoreManager *productRequestNotification = (StoreManager*)notification.object;
	PCSProductRequestStatus status = (PCSProductRequestStatus)productRequestNotification.status;

	if (status == PCSStoreResponse) {
		// Switch to the Products view controller.
		[self switchToViewController:ParentViewControllerSegmentProducts];
		[self.products reloadWithData:productRequestNotification.storeResponse];
		self.segmentedControl.selectedSegmentIndex = ParentViewControllerSegmentProducts;
	} else if (status == PCSRequestFailed) {
		[self alertWithTitle:PCSMessagesProductRequestStatus message:productRequestNotification.message];
	}
}

#pragma mark - Handle PCSPurchase Notification

/// Updates the UI according to the purchase request notification result.
-(void)handlePurchaseNotification:(NSNotification *)notification {
	StoreObserver *purchasesNotification = (StoreObserver *)notification.object;
	PCSPurchaseStatus status = (PCSPurchaseStatus)purchasesNotification.status;

	switch (status) {
		case PCSNoRestorablePurchases:
		case PCSPurchaseFailed:
		case PCSRestoreFailed: [self alertWithTitle:PCSMessagesPurchaseStatus message:purchasesNotification.message];
			break;
		// Switch to the Purchases view when receiving a successful restore notification.
		case PCSRestoreSucceeded: [self handleRestoredSucceededTransaction];
			break;
		default: break;
	}
}

#pragma mark - Handle Restored Transactions

/// Handles succesful restored transactions. Switches to the Purchases view.
-(void)handleRestoredSucceededTransaction {
	self.utility.restoreWasCalled = YES;

	// Refresh the Purchases view with the restored purchases.
	[self switchToViewController:ParentViewControllerSegmentPurchases];
	[self.purchases reloadWithData:self.utility.dataSourceForPurchasesUI];
	self.segmentedControl.selectedSegmentIndex = ParentViewControllerSegmentPurchases;
}

#pragma mark - Memory Management

- (void)dealloc {
	// Unregister for StoreManager's notifications.
	[[NSNotificationCenter defaultCenter] removeObserver:self name:PCSProductRequestNotification object:[StoreManager sharedInstance]];

	// Unregister for StoreObserver's notifications.
	[[NSNotificationCenter defaultCenter] removeObserver:self name:PCSPurchaseNotification object:[StoreObserver sharedInstance]];
}

@end

File: Swift/iOS/ParentViewController.swift
/*
See LICENSE folder for this sample’s licensing information.

Abstract:
Manages the child view controllers: Products and Purchases. Displays a Restore button that allows you to restore all previously purchased
non-consumable and auto-renewable subscription products. Requests product information about a list of product identifiers using StoreManager. Calls
StoreObserver to implement the restoration of purchases.
*/

import UIKit

class ParentViewController: UIViewController {
	// MARK: - Types

	fileprivate enum Segment: Int {
		case products, purchases
	}

	// MARK: - Properties

	@IBOutlet weak fileprivate var containerView: UIView!
	@IBOutlet weak fileprivate var segmentedControl: UISegmentedControl!
	@IBOutlet weak fileprivate var restoreButton: UIBarButtonItem!

	fileprivate var utility = Utilities()

	fileprivate lazy var products: Products = {
		let identifier = ViewControllerIdentifiers.products
		guard let controller = storyboard?.instantiateViewController(withIdentifier: identifier) as? Products
			else { fatalError("\(Messages.unableToInstantiateProducts)") }
		return controller
	}()

	fileprivate lazy var purchases: Purchases = {
		let identifier = ViewControllerIdentifiers.purchases
		guard let controller = storyboard?.instantiateViewController(withIdentifier: identifier) as? Purchases
			else { fatalError("\(Messages.unableToInstantiatePurchases)") }
		return controller
	}()

	// MARK: - View Life Cycle

	override func viewDidLoad() {
		super.viewDidLoad()

		// Disable or hide item.
		restoreButton.disable()

		StoreManager.shared.delegate = self
		StoreObserver.shared.delegate = self

		// Fetch product information.
		fetchProductInformation()
	}

	// MARK: - Switching Between View Controllers

	/// Adds a child view controller to the container.
	fileprivate func addBaseViewController(_ viewController: BaseViewController) {
		addChild(viewController)
		viewController.view.translatesAutoresizingMaskIntoConstraints = false
		viewController.view.frame = containerView.bounds
		containerView.addSubview(viewController.view)

		NSLayoutConstraint.activate([viewController.view.topAnchor.constraint(equalTo: containerView.safeAreaLayoutGuide.topAnchor),
									 viewController.view.bottomAnchor.constraint(equalTo: containerView.safeAreaLayoutGuide.bottomAnchor),
									 viewController.view.leadingAnchor.constraint(equalTo: containerView.safeAreaLayoutGuide.leadingAnchor),
									 viewController.view.trailingAnchor.constraint(equalTo: containerView.safeAreaLayoutGuide.trailingAnchor)])
		viewController.didMove(toParent: self)
	}

	/// Removes a child view controller from the container.
	fileprivate func removeBaseViewController(_ viewController: BaseViewController) {
		viewController.willMove(toParent: nil)
		viewController.view.removeFromSuperview()
		viewController.removeFromParent()
	}

	/// Switches between the Products and Purchases view controllers.
	fileprivate func switchToViewController(segment: Segment) {
		switch segment {
		case .products:
			removeBaseViewController(purchases)
			addBaseViewController(products)
		case .purchases:
			removeBaseViewController(products)
			addBaseViewController(purchases)
		}
	}

	// MARK: - Fetch Product Information

	/// Retrieves product information from the App Store.
	fileprivate func fetchProductInformation() {
		// First, let's check whether the user is allowed to make purchases. Proceed if they are allowed. Display an alert, otherwise.
		if StoreObserver.shared.isAuthorizedForPayments {
			restoreButton.enable()

			let resourceFile = ProductIdentifiers()
			guard let identifiers = resourceFile.identifiers else {
				// Warn the user that the resource file could not be found.
				alert(with: Messages.status, message: resourceFile.wasNotFound)
				return
			}

			if !identifiers.isEmpty {
				switchToViewController(segment: .products)
				// Refresh the UI with identifiers to be queried.
				products.reload(with: [Section(type: .invalidProductIdentifiers, elements: identifiers)])

				// Fetch product information.
				StoreManager.shared.startProductRequest(with: identifiers)
			} else {
				// Warn the user that the resource file does not contain anything.
				alert(with: Messages.status, message: resourceFile.isEmpty)
			}
		} else {
			// Warn the user that they are not allowed to make purchases.
			alert(with: Messages.status, message: Messages.cannotMakePayments)
		}
	}

	// MARK: - Display Alert

	/// Creates and displays an alert.
	fileprivate func alert(with title: String, message: String) {
		let alertController = utility.alert(title, message: message)
		self.navigationController?.present(alertController, animated: true, completion: nil)
	}

	// MARK: - Restore All Appropriate Purchases

	/// Called when tapping the "Restore" button in the UI.
	@IBAction func restore(_ sender: UIBarButtonItem) {
		// Calls StoreObserver to restore all restorable purchases.
		StoreObserver.shared.restore()
	}

	// MARK: - Handle Segmented Control Tap

	/// Called when tapping any segmented control in the UI.
	@IBAction func segmentedControlSelectionDidChange(_ sender: UISegmentedControl) {
		guard let segment = Segment(rawValue: sender.selectedSegmentIndex)
			else { fatalError("\(Messages.unknownSelectedSegmentIndex)\(sender.selectedSegmentIndex)).") }

		switchToViewController(segment: segment)

		switch segment {
		case .products: fetchProductInformation()
		case .purchases: purchases.reload(with: utility.dataSourceForPurchasesUI)
		}
	}

	// MARK: - Handle Restored Transactions

	/// Handles successful restored transactions. Switches to the Purchases view.
	fileprivate func handleRestoredSucceededTransaction() {
		utility.restoreWasCalled = true

		// Refresh the Purchases view with the restored purchases.
		switchToViewController(segment: .purchases)
		purchases.reload(with: utility.dataSourceForPurchasesUI)
		segmentedControl.selectedSegmentIndex = 1
	}
}

// MARK: - StoreManagerDelegate

/// Extends ParentViewController to conform to StoreManagerDelegate.
extension ParentViewController: StoreManagerDelegate {
	func storeManagerDidReceiveResponse(_ response: [Section]) {
		switchToViewController(segment: .products)
		// Switch to the Products view controller.
		products.reload(with: response)
		segmentedControl.selectedSegmentIndex = 0
	}

	func storeManagerDidReceiveMessage(_ message: String) {
		alert(with: Messages.productRequestStatus, message: message)
	}
}

// MARK: - StoreObserverDelegate

/// Extends ParentViewController to conform to StoreObserverDelegate.
extension ParentViewController: StoreObserverDelegate {
	func storeObserverDidReceiveMessage(_ message: String) {
		alert(with: Messages.purchaseStatus, message: message)
	}

	func storeObserverRestoreDidSucceed() {
		handleRestoredSucceededTransaction()
	}
}
